Bottom: 676b2a37e1a92dfa92e3d0c51cab14a68af0544b
Top:    046c980d12136b01884720fdf78b87972ba6633c
Author: Andrew Cooper <andrew.cooper@novatechweb.com>
Date:   2018-07-11 17:03:10 -0500

Configure Docker Daemon


---
diff --git a/ansible-playbook/group_vars/docker-hosts/docker.yml b/ansible-playbook/group_vars/docker-hosts/docker.yml
index 0d394562..4623feba 100644
--- a/ansible-playbook/group_vars/docker-hosts/docker.yml
+++ b/ansible-playbook/group_vars/docker-hosts/docker.yml
@@ -24,3 +24,5 @@ docker_network_frontend: 'frontend'
 
 # docker force build images
 docker_image_force_build: true
+
+docker_storage_driver: overlay2
diff --git a/ansible-playbook/host_vars/testdaedalus.novatech-llc.com/docker.yml b/ansible-playbook/host_vars/testdaedalus.novatech-llc.com/docker.yml
new file mode 100644
index 00000000..df7bb7de
--- /dev/null
+++ b/ansible-playbook/host_vars/testdaedalus.novatech-llc.com/docker.yml
@@ -0,0 +1,3 @@
+---
+
+docker_storage_driver: devicemapper
diff --git a/roles/docker-common/handlers/main.yml b/roles/docker-common/handlers/main.yml
new file mode 100644
index 00000000..8c329de7
--- /dev/null
+++ b/roles/docker-common/handlers/main.yml
@@ -0,0 +1,8 @@
+---
+- name: Restart Docker Daemon
+  systemd:
+    name: docker.service
+    daemon_reload: yes
+    enabled: yes
+    state: restarted
+  listen: "restart docker service"
diff --git a/roles/docker-common/tasks/configure.yml b/roles/docker-common/tasks/configure.yml
new file mode 100644
index 00000000..34754342
--- /dev/null
+++ b/roles/docker-common/tasks/configure.yml
@@ -0,0 +1,59 @@
+---
+
+- name: Create Docker service configuration directory
+  file:
+    path: /etc/systemd/system/docker.service.d
+    state: directory
+    owner: root
+    group: root
+
+- name: Configure Docker Daemon (host options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/00-host.conf
+    src: service-host.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  notify: "restart docker service"
+
+- name: Configure Docker Daemon (log driver options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/10-log.conf
+    src: service-log-json.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  notify: "restart docker service"
+
+- assert:
+    that: docker_thinpool_lv is defined
+  when: docker_storage_driver == "devicemapper"
+
+- name: Configure Docker Daemon (storage driver options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/20-storage.conf
+    src: service-storage-dm.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  when: docker_storage_driver == "devicemapper"
+  notify: "restart docker service"
+
+- name: Configure Docker Daemon (storage driver options)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/20-storage.conf
+    src: service-storage-ov2.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  when: docker_storage_driver == "overlay2"
+  notify: "restart docker service"
+
+- name: Configure Docker Daemon (service)
+  template: 
+    dest: /etc/systemd/system/docker.service.d/99-service.conf
+    src: service.conf.j2
+    owner: root
+    group: root
+    mode: 'u=rw,g=r,o=r'
+  notify: "restart docker service"
diff --git a/roles/docker-common/tasks/main.yml b/roles/docker-common/tasks/main.yml
index 7ef3f40c..bc8aac33 100644
--- a/roles/docker-common/tasks/main.yml
+++ b/roles/docker-common/tasks/main.yml
@@ -8,6 +8,7 @@
 - import_tasks: sshd_settings.yml
 - import_tasks: mariadb_settings.yml
 
+- import_tasks: configure.yml
 
 
 - name: Check if baculavirt selinux module exists
diff --git a/roles/docker-common/templates/daemon.json.j2 b/roles/docker-common/templates/daemon.json.j2
new file mode 100644
index 00000000..57de7390
--- /dev/null
+++ b/roles/docker-common/templates/daemon.json.j2
@@ -0,0 +1,4 @@
+{
+  "log-driver": "json-file",
+  "log-opts": {"max-size": "10m", "max-file": "3"}
+}
diff --git a/roles/docker-common/templates/service-host.conf.j2 b/roles/docker-common/templates/service-host.conf.j2
new file mode 100644
index 00000000..38c9f6e3
--- /dev/null
+++ b/roles/docker-common/templates/service-host.conf.j2
@@ -0,0 +1,5 @@
+[Service]
+Environment="HOST_OPTIONS=\
+--host=unix:///var/run/docker.sock \
+--host=tcp://{{ ansible_default_ipv4['address'] }}:2375 \
+"
diff --git a/roles/docker-common/templates/service-log-json.conf.j2 b/roles/docker-common/templates/service-log-json.conf.j2
new file mode 100644
index 00000000..a268ff76
--- /dev/null
+++ b/roles/docker-common/templates/service-log-json.conf.j2
@@ -0,0 +1,6 @@
+[Service]
+Environment="LOG_OPTIONS=\
+--log-driver=json-file \
+--log-opt max-size=10m \
+--log-opt max-file=3 \
+"
diff --git a/roles/docker-common/templates/service-storage-dm.conf.j2 b/roles/docker-common/templates/service-storage-dm.conf.j2
new file mode 100644
index 00000000..45532f6c
--- /dev/null
+++ b/roles/docker-common/templates/service-storage-dm.conf.j2
@@ -0,0 +1,8 @@
+[Service]
+Environment="STORAGE_OPTIONS=\
+--storage-driver devicemapper \
+--storage-opt dm.fs=xfs \
+--storage-opt dm.thinpooldev={{ docker_thinpool_lv.lv_dm_path }} \
+--storage-opt dm.use_deferred_removal=true \
+--storage-opt dm.use_deferred_deletion=true \
+"
diff --git a/roles/docker-common/templates/service-storage-ov2.conf.j2 b/roles/docker-common/templates/service-storage-ov2.conf.j2
new file mode 100644
index 00000000..3fa80a6e
--- /dev/null
+++ b/roles/docker-common/templates/service-storage-ov2.conf.j2
@@ -0,0 +1,4 @@
+[Service]
+Environment="STORAGE_OPTIONS=\
+--storage-driver overlay2 \
+"
diff --git a/roles/docker-common/templates/service.conf.j2 b/roles/docker-common/templates/service.conf.j2
new file mode 100644
index 00000000..7e552e00
--- /dev/null
+++ b/roles/docker-common/templates/service.conf.j2
@@ -0,0 +1,6 @@
+[Service]
+ExecStart=
+ExecStart=/usr/bin/dockerd \
+          $HOST_OPTIONS \
+          $LOG_OPTIONS \
+          $STORAGE_OPTIONS
